added rpartition

sohit-nayak opened the pull request:
added rpartition and tests.

freakboy3742 commented on tests/datatypes/test_str.py:
```diff
@@ -574,6 +574,18 @@ def test_rjust(self):
 
+    def test_rpartition(self):
+        self.assertCodeExecution("""
+            Stri = "Hello World!"
```
flake8 won't catch the problems with test code, but your test code should be flake8 compliant, too. Variable names shouldn't start with a capital letter.

freakboy3742 commented on tests/datatypes/test_str.py:
```diff
@@ -574,6 +574,18 @@ def test_rjust(self):
+            try:
+                print(Stri.rpartition(""))
+            except ValueError as err:
+                print(err)
```
This is a good start to testing, but there's a bunch of other tests needed - most notably, partitioning on a string that doesn't exist, and partitioning on a multi-character string.

sohit-nayak commented:
@freakboy3742 I added some more test cases, and found out that the tailing cases were'nt handled properly. I have modified the code, and guess its good to go now! :smile:

cflee commented on tests/datatypes/test_str.py:
```diff
@@ -574,6 +574,23 @@ def test_rjust(self):
+            print(st.rpartition("ld!"))
+            print("voc".rpartition("foobar"))
+            try:
+                print(st.rpartition(""))
```
It's great that an empty string is being tested here, but could you also test the cases of no parameter being passed (`rpartition()`), and passing in some non-string types?

It might also be worth testing whether it accepts an object that implements `__str__()`, to be thorough.

Also, you should test the case where `st = ""` as well.

cflee commented on python/common/org/python/types/Str.java:
```diff
@@ -1190,10 +1190,46 @@ private long toPositiveIndex(long index) {
+        if (sepStr.equals("")) {
+            throw new org.python.exceptions.ValueError("empty separator");
+        }
+        java.lang.String[] split = this.value.split(sepStr);
```
What happens if you pass in some separator string that can be interpreted as a regex? `java.lang.String.split()` will interpret it as a regex, but will python's `Str.rpartition()`?

If it shouldn't be interpreted as a regex, could you use something like `java.lang.String.lastIndexOf()` instead?

sohit-nayak replied:
hmm, I too think use of split is wrong here since it accepts regex, but rpartition doesnt.

cflee commented on python/common/org/python/types/Str.java:
```diff
@@ -1190,10 +1190,46 @@ private long toPositiveIndex(long index) {
+                str_one += split[i];
+                break;
+            }
+            str_one += split[i] + sepStr;
```
This loop seems rather inefficient. If the string is split into many parts (due to having many separator inside), each part needs to be individually concatenated to form the string. One way of making it more efficient would be to find the length of the part-after-the-separator and just do a `java.lang.String.substring()`.

cflee commented on python/common/org/python/types/Str.java:
```diff
@@ -1190,10 +1190,46 @@ private long toPositiveIndex(long index) {
+            return new org.python.types.Tuple(tuple);
+        }
+        if (split.length == 1) {
+            if ((this.value.substring(this.value.length() - sepStr.length())).equals(sepStr)) {
```
This condition seems a bit strange. In which case will it be true? It seems like you're taking the string of the last `sepStr.length()` chars and comparing it with `sepStr` itself?

sohit-nayak commented:
@cflee I have opened a fresh PR #512 with the requested changes. Please have a look. Thanks!
